File: Algorithms/110. Balanced Binary Tree/Solution.py
#coding=utf-8
__author__ = 'xuxuan'
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        return True if root is None else abs(self.maxDeep(root.left)-self.maxDeep(root.right))<=1 and self.isBalanced(root.left) and self.isBalanced(root.right)

    def maxDeep(self,root):
        return 0 if root is None else max(self.maxDeep(root.left),self.maxDeep(root.right))+1

File: Algorithms/110. Balanced Binary Tree/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean isBalanced(TreeNode root) {
        if (root==null) {
			return true;
		}
        if (Math.abs(f(root.left)-f(root.right))>1) {
			return false;
		}
        return isBalanced(root.left)&&isBalanced(root.right);
    }
    public int f(TreeNode root){
    	if (root==null) {
			return 0;
		}else {
			return 1+Math.max(f(root.left), f(root.right));
		}
    }
}